File: rust/2024/src/advent_utils.rs
use std::path::Path;
type DaySolution = (fn(&str) -> usize, fn(&str) -> usize);

pub trait Solution {
    fn part_1(data: &str) -> usize;
    fn part_2(data: &str) -> usize;
}

pub struct Solver {
    i: usize,
    solutions: [DaySolution; 25],
    path: &'static Path,
}

impl Solver {
    pub fn from(path: &'static str) -> Self {
        Self {
            i: 0,
            solutions: [(|_| 0, |_| 0); 25],
            path: path.as_ref(),
        }
    }
    pub fn add<S: Solution>(&self) -> Self {
        let mut s = Self { ..*self };
        s.solutions[s.i] = (S::part_1, S::part_2);
        s.i += 1;
        s
    }

    pub fn compute(self) {
        for (n, (part_1, part_2)) in self.solutions.into_iter().enumerate() {
            let day = format!("{:02}", n + 1);
            let fname = format!("day{day}.txt");
            let path = self.path.join(fname);
            run(path, day, part_1, part_2);
        }
    }
}
pub fn run(
    path: std::path::PathBuf,
    day: String,
    part_1: fn(&str) -> usize,
    part_2: fn(&str) -> usize,
) {
    if !path.exists() {
        return;
    }
    println!("Day {day} =>");
    let data = std::fs::read_to_string(path).expect("Error while reading the file. Aborting.");
    let data = data.trim();

    let result_1 = part_1(data);
    println!("\tPart 1: {result_1}");

    let result_2 = part_2(data);
    println!("\tPart 2: {result_2}\n");
}

File: zig/src/2024/advent_utils.zig
const std = @import("std");

pub const Solution = struct {
    part1: *const fn ([]u8) anyerror!i32,
    part2: *const fn ([]u8) anyerror!i32,
};

const Solver = struct {
    path: []const u8,
    solutions: [25]?Solution,
    idx: usize = 0,

    pub fn add(solver: *const Solver, sol: Solution) Solver {
        var s = solver.*;
        s.solutions[solver.idx] = sol;
        s.idx += 1;
        return s;
    }

    pub fn run(solver: Solver) anyerror!void {
        const writer = std.io.getStdOut().writer();
        var gp_alloc = std.heap.GeneralPurposeAllocator(.{}){};
        const gpa = gp_alloc.allocator();

        for (0..25) |i| {
            if (solver.solutions[i] == null) {
                continue;
            }
            const sol = solver.solutions[i].?;
            var buf: [2]u8 = undefined;
            var day: []u8 = undefined;

            if (i < 10) {
                day = try std.fmt.bufPrint(&buf, "0{d}", .{i + 1});
            } else {
                day = try std.fmt.bufPrint(&buf, "{d}", .{i + 1});
            }
            const path = try std.fmt.allocPrint(gpa, "{s}day{s}.txt", .{ solver.path, day });
            const data = try getInput(path);
            try writer.print("Day {s} =>\n", .{day});
            const res1 = try sol.part1(data);
            try writer.print("\tPart 1: {d}\n", .{res1});
            const res2 = try sol.part2(data);
            try writer.print("\tPart 2: {d}\n\n", .{res2});
        }
    }
};

fn getInput(path: []const u8) ![]u8 {
    const file = try std.fs.cwd().openFile(path, .{});
    defer file.close();

    const stat = try file.stat();
    const file_size = stat.size;
    var gp_alloc = std.heap.GeneralPurposeAllocator(.{}){};
    const gpa = gp_alloc.allocator();

    return try std.fs.cwd().readFileAlloc(gpa, path, file_size);
}

pub fn Runner(comptime path: []const u8) Solver {
    return Solver{ .path = path, .solutions = [_](?Solution){null} ** 25 };
}
